### PR TITLE
Request for comments: python 3.6 shim

### PR DESCRIPTION
**Pull request by skabbes**
## tldr
shim written in python3.6 allows up to run python3.6 code. POC for concurrent handler is slow.

## Details
Compared to apex, Up _always_ needs to run the node shim (Since up-proxy is written in go).  However, this prevents native support for all lambda languages.

This _happens_ to work for almost all supported languages since their binaries are embedded into the `nodejs6.10` runtime.

This notably breaks with the python3.6 runtime, since the 'python' available and the `PYTHONPATH` are all in reference to python 2.7.

This PR does 2 things:

1. Demonstrates python3.6 support is possible with up (with no changes elsewhere)
1. Demonstrates a multithreaded capable shim (assume `handler` is called concurrently)

The code to support 1 is dramatically simpler than the current node.js shim, mostly because python comes built in with a line reader.

The code for 2 requires a little extra help from go-apex, to boomerang the ID back to us (since we can receive concurrent responses, and need to tell the difference between them). [apex-go.patch.txt](https://github.com/apex/up/files/1322241/apex-go.patch.txt)

The concurrent proof of concept works, however it _appears_ to be fairly slow, which I need to dig into.

## Suggestions for next steps
After getting some feedback from the community, my suggest to move forward would be to simplify this code back to the single threaded version, make it python 2 and 3 compatible, and deploy this proxy _only_ for python3.6 functions.

If the benchmarks show it is just as fast as the node.js shim, I think the extra simplicity is worth replacing the nodejs shim with this (however my intuition says it will be slower).

**if concurrency is desired from the shim at all** (it may not be)
I suggest throwing out all of the concurrent code pending a refactor of the (shim -> go) communication to vanilla http, gRPC or some sort of request-response IPC framework that has already been performance tuned (vs my hand-rolled producer-consumer model below).

### CODE REVIEW

**Comment by tj** [comment]
hmmm Lambda only processes with a concurrency of 1 anyway, so that part might be unnecessary. I think if we're switching shims the only thing currently worth it is the Python -> Go bindings for the improved cold start times, but IIRC they were using Docker for that, might be too much of a hack

**Comment by skabbes** [comment]
@tj I poked around a bit - and couldn't really figure out how to get python 3.6 to run in the current up.  Maybe that's not a goal of the project - which is fine, but possibly just adding an [example](https://github.com/apex/up-examples) for python 3.6 would completely solve the problem.

**Comment by tj** [comment]
@skabbes ahhh I thought someone found that it was installed as `python36` or something? Maybe I misread that in Slack

**Comment by skabbes** [comment]
Going to close this, and re-open simplified and with new lambda research.

**Comment by hankyates** [comment]
So is the command for getting up to run in python 3.6:

``` 
"proxy": {
    "command": "FLASK_APP=app.py python36 -m flask run -p $PORT"
  }
```
